src/main/java/org/folio/rest/impl/ProtectionHelper.java

```diff
@@ -0,0 +1,117 @@
+
+
+  public ProtectionHelper(Map<String, String> okapiHeaders, Context ctx, String lang, ProtectedOperationType operation) {
+    super(okapiHeaders, ctx, lang);
+    acquisitionsUnitsHelper = new AcquisitionsUnitsHelper(okapiHeaders, ctx, lang);
+    acquisitionsUnitAssignmentsHelper = new AcquisitionsUnitAssignmentsHelper(okapiHeaders, ctx, lang);
+    this.operation = operation;
+  }
+
+  /**
+   * This method determines status of operation restriction based on ID of {@link CompositePurchaseOrder}.
+   * @param recordId corresponding record ID.
+   *
+   * @return true if operation is restricted, otherwise - false.
+   */
+  public CompletableFuture<Void> isOperationRestricted(String recordId) {
+    return getUnitIdsAssignedToOrder(recordId)
+      .thenCompose(this::isOperationRestricted);
+  }
+
+  /**
+   * This method determines status of operation restriction based on unit IDs from {@link CompositePurchaseOrder}.
+   * @param unitIds list of unit IDs.
+   *
+   * @return true if operation is restricted, otherwise - false.
+   */
+  public CompletableFuture<Void> isOperationRestricted(List<String> unitIds) {
+    if(!unitIds.isEmpty()) {
+      return getUnitsByIds(unitIds)
+        .thenCompose(units -> {
+          if(unitIds.size() == units.size()) {
+                if(applyMergingStrategy(units)) {
+                  return getUnitIdsAssignedToUserAndOrder(okapiHeaders.get(OKAPI_USERID_HEADER), unitIds)
+                    .thenAccept(ids -> {
+                      if(ids == 0) {
+                        throw new HttpException(HttpStatus.HTTP_FORBIDDEN.toInt(), USER_HAS_NOT_PERMISSIONS);
+                      }
+                    });
+                }
+            return CompletableFuture.completedFuture(null);
+          } else {
+            throw new HttpException(HttpStatus.HTTP_VALIDATION_ERROR.toInt(), ORDER_UNITS_NOT_FOUND);
+          }
+        });
+    } else {
+      return CompletableFuture.completedFuture(null);
+    }
+  }
+
+  /**
+   * This method checks existence of units associated with order.
+   * @param recordId id of order.
+   *
+   * @return true if units exist, otherwise - false.
+   */
+  private CompletableFuture<List<String>> getUnitIdsAssignedToOrder(String recordId) {
+    return acquisitionsUnitAssignmentsHelper.getAcquisitionsUnitAssignments(String.format("recordId==%s", recordId), 0, Integer.MAX_VALUE)
+      .thenApply(assignment -> assignment.getAcquisitionsUnitAssignments().stream().map(AcquisitionsUnitAssignment::getAcquisitionsUnitId).collect(toList()));
+  }
+
+  /**
+   * This method returns list of units ids associated with User.
+   * @param userId id of User.
+   *
+   * @return list of unit ids associated with user.
+   */
+  private CompletableFuture<Integer> getUnitIdsAssignedToUserAndOrder(String userId, List<String> unitIdsAssignedToOrder) {
+    String query = String.format("userId==%s AND %s", userId, convertIdsToCqlQuery(unitIdsAssignedToOrder, "acquisitionsUnitId"));
```

Comment:
This method is not checking if the user is part of all the acquisition units which have protectCreate=true:For example:
If MAIN and LAW both have protectCreate = true
User Bob is part of Only MAIN.
Order has acquistions Units : MAIN, LAW

This method returns the id of just MAIN, and an Order will be created with unit LAW even though Bob is not part of it and it has protect create

Reply:
@KVupp, according to the algorithm: https://wiki.folio.org/display/FOLIJET/Acquisitions+Units?preview=/25724746/25725391/Orders_Create.jpg#AcquisitionsUnits-Users. If order creation is protected by all order’s units we just need to check if User is member of order’s units.

What does this means?

Initially I supposed that we just need to check if User belongs to even one Unit from Order’s units. This case we don’t need to validate User’s protection status based on its units (included belonging and not belonging order’s units).

As per your comments this is not correct. And we need to check User’s protection status based on all assigned to it Units (even if part of Units is not belonging Order’s units). After that we can make decision if current operation is restricted or no. Am I correct?

Reply:
My understanding is User must not be able to create an order in acq unit that he doesn't belong to. @craigmcnally please confirm!

Reply:
if protectCreate == true, the user needs to belong to at least one of the specified acquisitions units.  I don't think we need to restrict them to being a member of all units the order is being created with.

This is consistent with the behavior of edit/update

Reply:
The [Acquisitions Units documentation](https://wiki.folio.org/display/FOLIJET/Acquisitions+Units) under "Conflicts and Priorities" states: "Bob, Ben, and Brenda all belong to at least one of the acquisitions units and can therefore perform any actions on this record."

Reply:
so essentially this comes down to, If I am the user, all I need is just to be part of a single Group with protect create = true, I have then the full power to create a record in any other acq unit that I wish(even though the creation is restricted to certain members).I still think it presents problems to the consortium colleges, however we can handle it when we get there.